### PR TITLE
Triggers

### PR DESCRIPTION
**Pull request by jdoliner**

### CODE REVIEW

File: triggers.md [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+// multiple triggers because the specs trigger atomically.
+message TriggerInfo {
+    Trigger trigger = 1;
+    repeated TriggerPair spec = 2;
+}
+```
+
+One thing to notice, a trigger is defined by a `TriggerPair` and it does not
+allow you to have a branch trigger another branch on a different repo. Why we
+need this constraint should be obvious but the fact that it's enforced by the
```
Comment (JoeyZwicker) [inline triggers.md]: Even after your explanation it's still not obvious to us lay people. can you elaborate further? Is this solely an implementation detail?

File: triggers.md [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+- cron: trigger a pipeline if a cron spec is satisfied and new data has been committed.
+- size: trigger a pipeline if an input has grown by a given amount.
+- commits: trigger a pipeline if input has received a given number of commits.
+
+The only trigger that might have made this list but didn't is number of files.
+This wouldn't be terribly hard to add, the only downside is that it requires
+walking the files which would make it less performant than the other triggers.
+
+It is not possible for a trigger to set more than one of these aspects.
```
Comment (JoeyZwicker) [inline triggers.md:27]: Can you explain why not? What if a user wants to trigger every 10MB or every 5 mins with 10MB hasn't come in by then? Sam's wants her deferred processes to do exactly this. As stated, that may be in the "do it yourself with branches" type of situation, but I'd like to understand what about this design specifically prevents someone from maybe wanting to do nested triggers like this?

File: triggers.md (hunk 1 above)
Comment (jdoliner) [reply triggers.md]: A trigger updates one branch to have the same head commit as another branch. That commit has to exist in a certain repo and both branches have to be on that same repo to point to it. It's similar to the constraint that you'll never get a branch that's in one repo with a head commit in a different repo.

File: triggers.md (hunk 2 above)
Comment (jdoliner) [reply triggers.md:27]: It's not hard to do this implementation wise, but it takes us down to the path to building a DSL. The obvious other ask if you can do 10M _or_ every 5 mins is can you do 10M _and_ every 5 mins? Both seem like they'd be pretty useful to me. I was trying to simplify as much as possible in the initial proposal, so I didn't include either one, but if you feel like they're necessary I'm fine adding them.

**Comment by JimmyWhitaker** [comment]
What do you think about having a "manual" trigger as well? I can imagine this being somewhat like a code or data review step where you would want a human "in the loop." It would essentially be implemented the same as adding a repo to the inputs with an empty commit to kick off a job without the overhead.

**Comment by jdoliner** [comment]
@JimmyWhitaker I'm on board with manual triggering it can already be accomplished with the command `pachctl create branch repo@triggered-branch --head repo@triggering-branch` that command sets the head of the triggered branch to be whatever the head of the triggering branch is, which is what happens when you trigger. That's admittedly not a super obvious command so I think adding another way to do the same thing as `pachctl run trigger` would be a nice addition. (I considered `pachctl trigger trigger` as well, but that one just reads too weirdly to me.)

**Pull request merged**